### ISSUE TITLE
1.9rc3 daemon cannot start on fresh RHEL7.1 vm in AWS

### ISSUE THREAD

**Issue by donhcd** [docker/docker#17464]
```
[ec2-user@ip-172-31-28-106 ~]$ sh
sh-4.2$ curl -sSL test.docker.com | sh
+ sudo -E sh -c 'sleep 3; yum -y -q install docker-engine'
warning: /var/cache/yum/x86_64/7Server/docker-testing-repo/packages/docker-engine-selinux-1.9.0-0.3.rc3.el7.centos.noarch.rpm: Header V4 RSA/SHA512 Signature, key ID 2c52609d: NOKEY
Public key for docker-engine-selinux-1.9.0-0.3.rc3.el7.centos.noarch.rpm is not installed
Importing GPG key 0x2C52609D:
 Userid     : "Docker Release Tool (releasedocker) <docker@docker.com>"
 Fingerprint: 5811 8e89 f3a9 1289 7c07 0adb f762 2157 2c52 609d
 From       : https://yum.dockerproject.org/gpg

If you would like to use Docker as a non-root user, you should now consider
adding your user to the "docker" group with something like:

  sudo usermod -aG docker ec2-user

Remember that you will have to log out and back in for this to take effect!

sh-4.2$ docker version
Client:
 Version:      1.9.0-rc3
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   2100b94
 Built:        Wed Oct 28 01:42:24 UTC 2015
 OS/Arch:      linux/amd64
Cannot connect to the Docker daemon. Is the docker daemon running on this host?
sh-4.2$ docker info
Cannot connect to the Docker daemon. Is the docker daemon running on this host?
sh-4.2$ sudo /bin/systemctl start docker.service
Job for docker.service failed. See 'systemctl status docker.service' and 'journalctl -xn' for details.
sh-4.2$ sudo systemctl status docker.service
docker.service - Docker Application Container Engine
   Loaded: loaded (/usr/lib/systemd/system/docker.service; disabled)
   Active: failed (Result: timeout) since Wed 2015-10-28 19:43:49 EDT; 1min 35s ago
     Docs: https://docs.docker.com
 Main PID: 10811

Oct 28 19:42:17 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: Starting Docker Application Container Engine...
Oct 28 19:42:17 ip-172-31-28-106.us-west-2.compute.internal docker[10811]: time="2015-10-28T19:42:17.912940792-04:00" level=info msg="API listen on /var/run/docker.sock"
Oct 28 19:42:17 ip-172-31-28-106.us-west-2.compute.internal docker[10811]: time="2015-10-28T19:42:17.997419286-04:00" level=warning msg="Usage of loopback devices is strongly discouraged for production use. Please use `--storage-opt dm.thinpooldev` or use `man docker` to r...hinpooldev section."
Oct 28 19:43:47 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: docker.service operation timed out. Terminating.
Oct 28 19:43:49 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: Failed to start Docker Application Container Engine.
Oct 28 19:43:49 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: Unit docker.service entered failed state.
Hint: Some lines were ellipsized, use -l to show in full.
sh-4.2$ sudo journalctl -xn
-- Logs begin at Wed 2015-10-28 19:34:55 EDT, end at Wed 2015-10-28 19:45:38 EDT. --
Oct 28 19:42:17 ip-172-31-28-106.us-west-2.compute.internal docker[10811]: time="2015-10-28T19:42:17.912940792-04:00" level=info msg="API listen on /var/run/docker.sock"
Oct 28 19:42:17 ip-172-31-28-106.us-west-2.compute.internal kernel: device-mapper: uevent: version 1.0.3
Oct 28 19:42:17 ip-172-31-28-106.us-west-2.compute.internal kernel: device-mapper: ioctl: 4.29.0-ioctl (2014-10-28) initialised: dm-devel@redhat.com
Oct 28 19:42:17 ip-172-31-28-106.us-west-2.compute.internal kernel: loop: module loaded
Oct 28 19:42:17 ip-172-31-28-106.us-west-2.compute.internal docker[10811]: time="2015-10-28T19:42:17.997419286-04:00" level=warning msg="Usage of loopback devices is strongly discouraged for production use. Please use `--storage-opt dm.thinpooldev` or use `man docker` to refer to dm.thinpooldev
Oct 28 19:43:47 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: docker.service operation timed out. Terminating.
Oct 28 19:43:49 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: Failed to start Docker Application Container Engine.
-- Subject: Unit docker.service has failed
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit docker.service has failed.
--
-- The result is failed.
Oct 28 19:43:49 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: Unit docker.service entered failed state.
Oct 28 19:45:24 ip-172-31-28-106.us-west-2.compute.internal sudo[10860]: ec2-user : TTY=pts/0 ; PWD=/home/ec2-user ; USER=root ; COMMAND=/bin/systemctl status docker.service
Oct 28 19:45:38 ip-172-31-28-106.us-west-2.compute.internal sudo[10863]: ec2-user : TTY=pts/0 ; PWD=/home/ec2-user ; USER=root ; COMMAND=/bin/journalctl -xn
sh-4.2$ uname -a
Linux ip-172-31-28-106.us-west-2.compute.internal 3.10.0-229.el7.x86_64 #1 SMP Thu Jan 29 18:37:38 EST 2015 x86_64 x86_64 x86_64 GNU/Linux
sh-4.2$ cat /etc/redhat-release
Red Hat Enterprise Linux Server release 7.1 (Maipo)
```

**Comment by donhcd** [comment]
more logs:
```
sh-4.2$ sudo journalctl -xeu docker
-- Logs begin at Wed 2015-10-28 19:34:55 EDT, end at Wed 2015-10-28 20:20:29 EDT. --
Oct 28 19:42:17 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: Starting Docker Application Container Engine...
-- Subject: Unit docker.service has begun with start-up
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit docker.service has begun starting up.
Oct 28 19:42:17 ip-172-31-28-106.us-west-2.compute.internal docker[10811]: time="2015-10-28T19:42:17.912940792-04:00" level=info msg="API listen on /var/run/docker.sock"
Oct 28 19:42:17 ip-172-31-28-106.us-west-2.compute.internal docker[10811]: time="2015-10-28T19:42:17.997419286-04:00" level=warning msg="Usage of loopback devices is strongly discouraged for production use. Please use `--storage-opt dm.thinpooldev` or use `man docker` to refer to dm.thinpooldev
Oct 28 19:43:47 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: docker.service operation timed out. Terminating.
Oct 28 19:43:49 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: Failed to start Docker Application Container Engine.
-- Subject: Unit docker.service has failed
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit docker.service has failed.
--
-- The result is failed.
Oct 28 19:43:49 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: Unit docker.service entered failed state.
Oct 28 20:08:29 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: Starting Docker Application Container Engine...
-- Subject: Unit docker.service has begun with start-up
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit docker.service has begun starting up.
Oct 28 20:08:29 ip-172-31-28-106.us-west-2.compute.internal docker[10954]: time="2015-10-28T20:08:29.482643955-04:00" level=info msg="API listen on /var/run/docker.sock"
Oct 28 20:08:29 ip-172-31-28-106.us-west-2.compute.internal docker[10954]: time="2015-10-28T20:08:29.511076798-04:00" level=warning msg="Usage of loopback devices is strongly discouraged for production use. Please use `--storage-opt dm.thinpooldev` or use `man docker` to refer to dm.thinpooldev
Oct 28 20:09:25 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: Stopping Docker Application Container Engine...
-- Subject: Unit docker.service has begun shutting down
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit docker.service has begun shutting down.
Oct 28 20:09:25 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: docker.service: main process exited, code=exited, status=2/INVALIDARGUMENT
Oct 28 20:09:25 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: Stopped Docker Application Container Engine.
-- Subject: Unit docker.service has finished shutting down
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit docker.service has finished shutting down.
Oct 28 20:09:25 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: Unit docker.service entered failed state.
Oct 28 20:15:29 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: Starting Docker Application Container Engine...
-- Subject: Unit docker.service has begun with start-up
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit docker.service has begun starting up.
Oct 28 20:15:30 ip-172-31-28-106.us-west-2.compute.internal docker[20858]: time="2015-10-28T20:15:30.038030041-04:00" level=info msg="API listen on /var/run/docker.sock"
Oct 28 20:15:30 ip-172-31-28-106.us-west-2.compute.internal docker[20858]: time="2015-10-28T20:15:30.040180033-04:00" level=warning msg="Usage of loopback devices is strongly discouraged for production use. Please use `--storage-opt dm.thinpooldev` or use `man docker` to refer to dm.thinpooldev
Oct 28 20:16:59 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: docker.service operation timed out. Terminating.
Oct 28 20:18:29 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: docker.service stopping timed out (2). Killing.
Oct 28 20:18:29 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: Failed to start Docker Application Container Engine.
-- Subject: Unit docker.service has failed
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit docker.service has failed.
--
-- The result is failed.
Oct 28 20:18:29 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: Unit docker.service entered failed state.
```

**Comment by jfrazelle** [comment]
ping @vbatts do you think devmapper issue or @mavenugo something w networking

**Comment by vbatts** [comment]
so we're looking at a RHEL7.1 host, that is pulling down the dockerproject.org rpm. It times-out when starting the daemon?

**Comment by vbatts** [comment]
The loopback warning doesn't fail launch presently. It may be worth a try to add `-s overlay` to the /etc/sysconfig/docker settings to try a different storage driver. Also `-D` for debug output.

**Comment by jfrazelle** [comment]
yeah the service seems to be hanging when starting?

**Comment by cpuguy83** [comment]
Yeah, wonder if it's devicemapper provisioning and it's taking too long?

**Comment by cpuguy83** [comment]
Maybe similar to #17448 if that is the case?

**Comment by donhcd** [comment]
after adding `-s overlay -D` it looks like things are working now. here's a link to the new logs from `sudo journalctl -xeu docker`: https://gist.github.com/donhcd/f6952a1f7203ee583ef9

**Comment by vbatts** [comment]
@donhcd thanks. for debugging sake, would you mind switching it back, but keeping the debug flag (`-s devicemapper -D`). This way we might gain more needed information to investigate the root cause.

**Comment by donhcd** [comment]
I changed the flags to `-s devicemapper -D` and ran `sudo systemctl stop docker && sudo systemctl start docker`. The start did not hang and appeared to succeed, but looks like the daemon immediately died after starting. Logs:

```
Oct 29 14:25:07 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: Stopping docker.service...
-- Subject: Unit docker.service has begun shutting down
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit docker.service has begun shutting down.
Oct 29 14:25:07 ip-172-31-28-106.us-west-2.compute.internal docker[28271]: time="2015-10-29T14:25:07.608901731-04:00" level=info msg="Processing signal 'terminated'"
Oct 29 14:25:07 ip-172-31-28-106.us-west-2.compute.internal docker[28271]: time="2015-10-29T14:25:07.609436177-04:00" level=debug msg="starting clean shutdown of all containers..."
Oct 29 14:25:07 ip-172-31-28-106.us-west-2.compute.internal docker[28271]: time="2015-10-29T14:25:07.609650088-04:00" level=debug msg="Cleaning up old shm/mqueue mounts: start."
Oct 29 14:25:07 ip-172-31-28-106.us-west-2.compute.internal docker[28271]: time="2015-10-29T14:25:07.609986268-04:00" level=debug msg="Cleaning up old shm/mqueue mounts: done."
Oct 29 14:25:07 ip-172-31-28-106.us-west-2.compute.internal docker[28271]: time="2015-10-29T14:25:07.610051527-04:00" level=debug msg="Unix socket /var/lib/docker/network/files/36ca1c2f4072c9f3a5987aeb3f4a86791b1e4d92ef488a0e0407d9a829f9a5e6.sock doesnt exist. cannot accept client connections"
Oct 29 14:25:07 ip-172-31-28-106.us-west-2.compute.internal docker[28271]: time="2015-10-29T14:25:07.610071317-04:00" level=debug msg="Clean shutdown succeeded"
Oct 29 14:25:07 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: Stopped docker.service.
-- Subject: Unit docker.service has finished shutting down
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit docker.service has finished shutting down.
Oct 29 14:25:10 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: Starting docker.service...
-- Subject: Unit docker.service has begun with start-up
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit docker.service has begun starting up.
Oct 29 14:25:10 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: Started docker.service.
-- Subject: Unit docker.service has finished start-up
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit docker.service has finished starting up.
--
-- The start-up result is done.
Oct 29 14:25:10 ip-172-31-28-106.us-west-2.compute.internal docker[29584]: time="2015-10-29T14:25:10.328885124-04:00" level=fatal msg="No sockets found"
Oct 29 14:25:10 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: docker.service: main process exited, code=exited, status=1/FAILURE
Oct 29 14:25:10 ip-172-31-28-106.us-west-2.compute.internal systemd[1]: Unit docker.service entered failed state.
```

I'll try this from a fresh install and see what I get then

**Comment by donhcd** [comment]
okay, from a fresh vm, I ran:

```
curl -sSL test.docker.com | sh
sudo sh -c 'echo -e "[Service]\nExecStart=/usr/bin/docker daemon -H fd:// -s devicemapper -D" > /etc/systemd/system/docker.service'
sudo systemctl daemon-reload
sudo systemctl start docker
sudo journalctl -xeu docker
```

the behavior was the same as above, and the logs returned by `journalctl` command seem to be the same:

```
Oct 29 14:34:51 ip-172-31-30-61.us-west-2.compute.internal systemd[1]: Starting docker.service...
-- Subject: Unit docker.service has begun with start-up
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit docker.service has begun starting up.
Oct 29 14:34:51 ip-172-31-30-61.us-west-2.compute.internal systemd[1]: Started docker.service.
-- Subject: Unit docker.service has finished start-up
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit docker.service has finished starting up.
--
-- The start-up result is done.
Oct 29 14:34:51 ip-172-31-30-61.us-west-2.compute.internal docker[10811]: time="2015-10-29T14:34:51.466842286-04:00" level=fatal msg="No sockets found"
Oct 29 14:34:51 ip-172-31-30-61.us-west-2.compute.internal systemd[1]: docker.service: main process exited, code=exited, status=1/FAILURE
Oct 29 14:34:51 ip-172-31-30-61.us-west-2.compute.internal systemd[1]: Unit docker.service entered failed state.
```

**Comment by donhcd** [comment]
I found #11135, so I tried removing the `-H fd://` flags (they were there by default) and that got rid of the "No sockets found" error. Now, the startup logs look like this:

```
Oct 29 17:16:13 ip-172-31-20-21.us-west-2.compute.internal systemd[1]: Starting docker.service...
-- Subject: Unit docker.service has begun with start-up
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit docker.service has begun starting up.
Oct 29 17:16:13 ip-172-31-20-21.us-west-2.compute.internal systemd[1]: Started docker.service.
-- Subject: Unit docker.service has finished start-up
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit docker.service has finished starting up.
--
-- The start-up result is done.
Oct 29 17:16:13 ip-172-31-20-21.us-west-2.compute.internal docker[11065]: time="2015-10-29T17:16:13.068913666-04:00" level=debug msg="docker group found. gid: 995"
Oct 29 17:16:13 ip-172-31-20-21.us-west-2.compute.internal docker[11065]: time="2015-10-29T17:16:13.069180650-04:00" level=debug msg="Server created for HTTP on unix (/var/run/docker.sock)"
Oct 29 17:16:13 ip-172-31-20-21.us-west-2.compute.internal docker[11065]: time="2015-10-29T17:16:13.069492212-04:00" level=debug msg="[graphdriver] trying provided driver \"devicemapper\""
Oct 29 17:16:13 ip-172-31-20-21.us-west-2.compute.internal docker[11065]: time="2015-10-29T17:16:13.071825091-04:00" level=info msg="API listen on /var/run/docker.sock"
Oct 29 17:16:13 ip-172-31-20-21.us-west-2.compute.internal docker[11065]: time="2015-10-29T17:16:13.092892749-04:00" level=debug msg="devicemapper: driver version is 4.29.0"
Oct 29 17:16:13 ip-172-31-20-21.us-west-2.compute.internal docker[11065]: time="2015-10-29T17:16:13.093386645-04:00" level=debug msg="Generated prefix: docker-202:2-8390534"
Oct 29 17:16:13 ip-172-31-20-21.us-west-2.compute.internal docker[11065]: time="2015-10-29T17:16:13.093408383-04:00" level=debug msg="devmapper: Checking for existence of the pool docker-202:2-8390534-pool"
Oct 29 17:16:13 ip-172-31-20-21.us-west-2.compute.internal docker[11065]: time="2015-10-29T17:16:13.094396466-04:00" level=debug msg="Pool doesn't exist. Creating it."
Oct 29 17:16:13 ip-172-31-20-21.us-west-2.compute.internal docker[11065]: time="2015-10-29T17:16:13.094467325-04:00" level=debug msg="Creating loopback file /var/lib/docker/devicemapper/devicemapper/data for device-manage use"
Oct 29 17:16:13 ip-172-31-20-21.us-west-2.compute.internal docker[11065]: time="2015-10-29T17:16:13.107812984-04:00" level=debug msg="Creating loopback file /var/lib/docker/devicemapper/devicemapper/metadata for device-manage use"
Oct 29 17:16:13 ip-172-31-20-21.us-west-2.compute.internal docker[11065]: time="2015-10-29T17:16:13.166436287-04:00" level=warning msg="Usage of loopback devices is strongly discouraged for production use. Please use `--storage-opt dm.thinpooldev` or use `man docker` to refer to dm.thinpooldev s
Oct 29 17:16:13 ip-172-31-20-21.us-west-2.compute.internal docker[11065]: time="2015-10-29T17:16:13.166507559-04:00" level=debug msg="Initializing base device-mapper thin volume"
Oct 29 17:16:13 ip-172-31-20-21.us-west-2.compute.internal docker[11065]: time="2015-10-29T17:16:13.169080704-04:00" level=debug msg="[devmapper] CreateDevice(poolName=/dev/mapper/docker-202:2-8390534-pool, deviceID=1)"
Oct 29 17:16:13 ip-172-31-20-21.us-west-2.compute.internal docker[11065]: time="2015-10-29T17:16:13.172155668-04:00" level=debug msg="Registering device (id 1) with FS size 107374182400"
Oct 29 17:16:13 ip-172-31-20-21.us-west-2.compute.internal docker[11065]: time="2015-10-29T17:16:13.172180963-04:00" level=debug msg="registerDevice(1, )"
Oct 29 17:16:13 ip-172-31-20-21.us-west-2.compute.internal docker[11065]: time="2015-10-29T17:16:13.174891269-04:00" level=debug msg="Creating filesystem on base device-mapper thin volume"
Oct 29 17:16:13 ip-172-31-20-21.us-west-2.compute.internal docker[11065]: time="2015-10-29T17:16:13.174911857-04:00" level=debug msg="activateDeviceIfNeeded()"
```

Then I tried some `docker run` commands, which all seemed to be hanging. I let one command run extra long and finally the engine pulled busybox and things worked mostly as expected:

```
[ec2-user@ip-172-31-20-21 ~]$ sudo docker run scratch
^C[ec2-user@ip-172-31-20-21 ~]$ sudo docker run -it scratch
^C[ec2-user@ip-172-31-20-21 ~]$ sudo docker run -it scratch bash
^C[ec2-user@ip-172-31-20-21 ~]$ sudo docker run -it busybox bash
Unable to find image 'busybox:latest' locally
latest: Pulling from library/busybox
ab2b8a86ca6c: Pull complete
2c5ac3f849df: Pull complete
Digest: sha256:5551dbdfc48d66734d0f01cafee0952cb6e8eeecd1e2492240bf2fd9640c2279
Status: Downloaded newer image for busybox:latest
exec: "bash": executable file not found in $PATH
Error response from daemon: Cannot start container 5b1ef600287a2c6837c327a25cfec909710f7bca786184394ef1231446522101: [8] System error: exec: "bash": executable file not found in $PATH
[ec2-user@ip-172-31-20-21 ~]$ sudo docker run scratch echo hi
Unable to find image 'scratch:latest' locally
Error response from daemon: 'scratch' is a reserved name
[ec2-user@ip-172-31-20-21 ~]$ sudo docker run busybox echo hi
hi
[ec2-user@ip-172-31-20-21 ~]$
```

coming back to the daemon logs, I saw these logs (I repeated the last line from the previous logs snippet to show the difference in time): https://gist.github.com/donhcd/5df51ae55eff6f1f8486

Why are there nearly 5 minutes between starting the daemon and the daemon actually being useful?

**Comment by cpuguy83** [comment]
Looks like devicemapper is the culprit in the long startup:

```

Oct 29 17:16:13 ip-172-31-20-21.us-west-2.compute.internal docker[11065]: time="2015-10-29T17:16:13.174911857-04:00" level=debug msg="activateDeviceIfNeeded()"
Oct 29 17:21:30 ip-172-31-20-21.us-west-2.compute.internal docker[11065]: time="2015-10-29T17:21:30.190972667-04:00" level=debug msg="activateDeviceIfNeeded()"
```

**Comment by rhvgoyal** [comment]
I think it is slow xfs initialization on 100G device for a thin pool which  is on top of loop devices. We have noticed the instances of systemd timing out and as a stop gap measure, increasing systemd timeout from 90 seconds to 5 minutes.

Can you try passing option `--storage-opt dm.basesize=10G` to daemon command line and see if problem is resolved.

**Comment by donhcd** [comment]
okay, so I can't repro the slow start on the same machine with `systemctl stop docker && systemctl start docker`, but I could when I spun up a new machine and repeated the steps. from a brand new vm, here's my history:

```
    1  export EDITOR=vi
    2
    3  curl -sSL test.docker.com | sh
    4  sudo sh -c 'echo -e "[Service]\nExecStart=/usr/bin/docker daemon -s devicemapper -D --storage-opt dm.basesize=10G" > /etc/systemd/system/docker.service'
    5  sudo systemctl daemon-reload
    6  sudo systemctl start docker
```

Right after, I ran a `docker run` and it worked as expected:
```
$ sudo docker run busybox
Unable to find image 'busybox:latest' locally
latest: Pulling from library/busybox
ab2b8a86ca6c: Pull complete
2c5ac3f849df: Pull complete
Digest: sha256:5551dbdfc48d66734d0f01cafee0952cb6e8eeecd1e2492240bf2fd9640c2279
Status: Downloaded newer image for busybox:latest
```

logs output from `sudo journalctl -xeu docker` [here](https://gist.github.com/donhcd/edaa47bddf503d84037d)

**Comment by mbentley** [comment]
I've seen the same and I see that `mkfs.ext4 -E nodiscard,lazy_itable_init=0,lazy_journal_init=0 /dev/mapper/docker-202:1-25558685-base` is running this CentOS 7 EC2 instance I am working on.

I could see that it was waiting on the `mkfs.ext4` process to complete when checking the status while waiting for `systemctl start docker` to complete:

```
# systemctl status docker -l
docker.service - Docker Application Container Engine
   Loaded: loaded (/usr/lib/systemd/system/docker.service; disabled)
   Active: deactivating (final-sigterm) (Result: timeout)
     Docs: https://docs.docker.com
 Main PID: 10752
   CGroup: /system.slice/docker.service
           └─10789 mkfs.ext4 -E nodiscard,lazy_itable_init=0,lazy_journal_init=0 /dev/mapper/docker-202:1-25558685-base

Oct 31 15:32:13 ip-172-31-30-107 systemd[1]: Starting Docker Application Container Engine...
Oct 31 15:32:13 ip-172-31-30-107 docker[10752]: time="2015-10-31T15:32:13.695345145Z" level=info msg="API listen on /var/run/docker.sock"
Oct 31 15:32:13 ip-172-31-30-107 docker[10752]: time="2015-10-31T15:32:13.920344399Z" level=warning msg="Usage of loopback devices is strongly discouraged for production use. Please use `--storage-opt dm.thinpooldev` or use `man docker` t
o refer to dm.thinpooldev section."
Oct 31 15:33:43 ip-172-31-30-107 systemd[1]: docker.service operation timed out. Terminating.
```

**Comment by runcom** [comment]
I'm seeing docker is taking forever to start up on Fedora Rawhide as well, and the process is actually hanging on `activateDeviceIfNeeded`

**Comment by tiborvass** [comment]
FWIW, I could reproduce not only with 1.9 but also with 1.8 :( So is there something that changed on AWS ? I could not reproduce this on DigitalOcean on Centos 7.1, neither with 1.8 nor 1.9.

**Comment by rhvgoyal** [comment]
I was thinking that problem is due to 100G device but looks like it has something to do with ext4. I changed my fs type to xfs and it seems to work well with both 10G and 100G base device size.

Those who are facing problems, can you try following as option to docker daemon.

`-s devicemapper --storage-opt dm.fs=xfs`

I just launched an fedora 22 instances in AWS and it is running docker 1.8 and by default ext4 is being used and systemd time out is happening. Even if I use dm.basesize=10G, it does not fix the problem for me. But if I force usage of xfs, it works just fine.

So this does not look like specific to docker 1.9.

**Comment by rhvgoyal** [comment]
@runcom from logs it looks like as if activateDeviceIfNeeded() is the problem but that's not the case. In this case somehow mkfs.ext4 takes a very long time and systemd times out. It is not clear though why does it take so long. May be slow cloud storage?

**Comment by tiborvass** [comment]
@rhvgoyal it worked for me

**Comment by tiborvass** [comment]
@rhvgoyal do you recommend this setting for all RHEL/CentOS setups or just the ones on AWS ?

**Comment by rhvgoyal** [comment]
@tiborvass we try to use xfs as default for all rhel/centos. docker-storage-setup enforces that.

https://github.com/projectatomic/docker-storage-setup/blob/master/docker-storage-setup.sh#L121

If docker-storage-setup fails because there is no free space to setup thin pool, then we are not able to override default fs (ext4) setting and that's where the problem with AWS seem to be visible.

I think it is a good idea to use xfs for all RHEL/CentOS.

**Comment by runcom** [comment]
@rhvgoyal forgot to say that the Fedora rawhide I tested with was a virtual machine if this could help further (virt install, virsh managed, qemu/kvm)

**Comment by rhvgoyal** [comment]
@runcom, I tested on a fedora 22 VM on my local machine both with ext4 and xfs  (on top of loop devices) an I can't reproduce the problem. So somehow problem is only visible with ext4 in AWS.

**Comment by runcom** [comment]
Mmm weird (vm's host is Fedora 22 on my laptop not AWS)

**Comment by rhvgoyal** [comment]
@runcom ok, so you have F22 host and rawhide VM and you are able to see the problem. Does using xfs as container fs make the problem go away?

**Comment by runcom** [comment]
I'll try to reproduce with xfs as the container fs and report here, thanks!

**Comment by rhvgoyal** [comment]
Ok, I removed options `lazy_itable_init=0` and `lazy_journal_init=0` and that fixes the problem for ext4 on AWS. docker initializes with-in 2 seconds.

So looks like it boils down to how much fs data gets created at mkfs time and how much of it is deferred. I am not sure why did we use these two options to begin with. What's the rationale. Why not stick with defaults.

**Comment by rhvgoyal** [comment]
So what's the way forward. I see two options.

- Use xfs by default. 

- Drop mkfs.ext4 options lazy_itable_init=0 and lazy_journal_init=0. This seems to force initializaiton of all metadata at creation time and it does not seem to be default. If we don't initialize all metadata at mkfs time and let a thread do it later after mount, then downside seems to be that we will be doing it on containers (snapshots) and if there are N container from 1 image, then we will do it N times and use much more thin pool space. So in case of ext4 trade off seems to be initialization time vs thin pool usage.

@sandeen mentioned that XFS does not have knobs for preallocating metadata. It dynamically allocates it as needed.

**Comment by tiborvass** [comment]
@rhvgoyal which option do you prefer? XFS ?

**Comment by rhvgoyal** [comment]
@tiborvass we have been using xfs as default and we enforce it in docker-storage-setup. So I feel xfs should be a good default choice. So I prefer making xfs as default.

**Comment by tiborvass** [comment]
@rhvgoyal ok thank you, since this is an AWS specific issue, I think it's less intrusive right now for 1.9 to not change the default in docker, and change our install scripts instead. I'll open a PR to close this issue.

**Comment by rhvgoyal** [comment]
@tiborvass That's fine.

**Comment by rhvgoyal** [comment]
@alexlarsson I noticed that from the beginning we have been using options 'lazy_itable_init=0,lazy_journal_init=0' for mkfs.ext4. I am wondering why did you choose to do it that way. Pre-allocating all the metadata on slower storage is turning out to be slow. 10G base image itself took more then 3 minutes only my system. Now default base image size is 100G and we have no chance to initialize it in time.

I ran some blktrace and on 10G image, ext4 did around 300MB of writes and I got around 1.5MB/s as overall throughput. I see lots of writes are synchronous writes and that might be responsible for this or may be IO pattern is fairly random. 

Anyway, I thought I will quickly ping you and understand what's your take on this issue.

**Comment by anusha-ragunathan** [comment]
https://github.com/docker/docker/issues/3280 has some background on using the lazy-* options. However, the thread later has an open ended discussion on why using lazy_itables_init might be poor use in this case. 
"<alex> i.e. although its likely to *work*, it will cause a lot of writes in each snapshot making them unnecessarily different"

Any comments on this? Thanks.

**Comment by jeremyeder** [comment]
@rhvgoyal I will forward you the internal thread wrt the lazy_init stuff.  It's 2 years old at this point.  Bottom line is that it's not terribly relevant these days -- your advice to use xfs is correct; which as you stated has been present in RHEL/Fedora packages for a long time.

Thank you @tiborvass for merging the change to xfs!

**Comment by tiborvass** [comment]
@rhvgoyal I used to be able to reproduce this, but now I can't. It took `28sec` to boot, but it worked just fine.

**Comment by tiborvass** [comment]
@jeremyeder @rhvgoyal do you guys think it's safe to remove `lazy_init=0` and `lazy_journald=0` for 1.9.1 ? We can try to make xfs default for 1.10.

**Comment by rhvgoyal** [comment]
@tiborvass This will happen only during first time docker is started and loop files are setup and file system is created. On subsequent restart of docker, base image is already there with filesystem on it so one would not face this issue.

**Comment by rhvgoyal** [comment]
@tiborvass I don't think getting rid of 'lazy_init=0' is an option. Reason being that would mean that we will  create ext4 metadata per container and that will lead to huge pool usage.

For example, with 100G thin device and with ext4 lazy_init=0, thin pool usage is around 1.8GB. That means there is roughly around 1.8GB of metadata which will be initialized per container if we disable pre-allocation.

I verified this by launching two containers and slowly pool usage goes up by around 1.5GB per container instance.

So I don't think that making 'lazy_init=0' is really an option.

**Comment by rhvgoyal** [comment]
@tiborvass So we had some conversations about ext4 behavior internally. @sandeen looked into a possible ext4 optimization where zeroing is skipped if discards are issued and device advertizes discards_zeros_data. Thin devices set discards_zeros_data=0 hence ext4 optimization does not kick in.

Now thought is that there does not seem to be a good way to support discards_zeros_data=1. Discards themselves are optional and blkdev_issue_discard() interface can chop off some of the discard area in the beginning and at the end. So current thought is that discards_zeros_data=1
is very hard and it seems to be contradictory to the notion of discards being optional.

In my testing AWS is not always slow. Today 100G thin device initialized in 1.5 minute. (as compared to taking 3 minutes on a 10G device). So looks like this varies a lot.

So either we switch to xfs by default (and fall back to ext4 if mkfs.xfs is not there). Or we just live with ext4 being default and if somebody runs into issues, suggest them to use dm.fs=xfs.

**Comment by tiborvass** [comment]
@rhvgoyal Have you tried on a fresh RHEL VM on AWS ? It not only takes 1m30, it fails as explained in the issue. Try `curl -SL https://get.docker.com/ | sh` on a fresh RHEL on AWS and try `sudo systemctl start docker` and you'll see the failure.

I think we'll go with the xfs by default if mkfs.xfs is present. We just need to make sure to test it on all OSes.

**Comment by rhvgoyal** [comment]
@tiborvaas I tried on Fedora 23 with a modified systemd docker.service unit file with timeout set to 5 minutes instead of 1 minute.

Point I was raising was that this mkfs.ext4 time seems to vary a lot on AWS. So sometimes it will fail and sometimes it will pass.

Apart from mfs.xfs being present I think we will have to make sure kernel supports xfs too. Otherwise we could have a case where mkfs.xfs is installed but kernel has been compiled with xfs support.

**Comment by tiborvass** [comment]
@rhvgoyal I only care about the default AMIs provided on AWS console, that's the most important right now since that's what people seem to spawn. If you have issues creating VMs on AWS for various reasons, can you give me your public ssh key so you I can add you to a fresh VM? Also would be easier to communicate on IRC or Slack. Let me know.

**Comment by tiborvass** [comment]
Closed by #17903
